### PR TITLE
feature[next]: Collect shifts for all nodes in TraceShift pass

### PR DESCRIPTION
**Pull request by tehrengruber**
- Shifts are collected not only for the closure inputs, but for every iterator expression in the tree (including iterator arguments to lambdas).
- The collected shifts are now represented as a set.

This PR is a prerequisite for the temporary extraction heuristics.

### CODE REVIEW

File: src/gt4py/next/iterator/transforms/trace_shifts.py [hunk 1]
```diff
@@ -26,6 +26,30 @@ class Sentinel(enum.Enum):
...
+class ShiftRecorder:
+    recorded_shifts: dict[int, set[tuple[ir.OffsetLiteral, ...]]] = dataclasses.field(
+        default_factory=dict
+    )
+
+    def register_node(self, inp: ir.Expr | ir.Sym) -> None:
+        self.recorded_shifts.setdefault(id(inp), set())
+
+    def __call__(self, inp: ir.Expr | ir.Sym, offsets: tuple[ir.OffsetLiteral, ...]) -> None:
+        self.recorded_shifts[id(inp)].add(offsets)
```
Comment (havogt) [inline src/gt4py/next/iterator/transforms/trace_shifts.py:39]:
```suggestion
    def __call__(self, inp: ir.Expr | ir.Sym, offsets: tuple[ir.OffsetLiteral, ...]) -> None:
        self.recorded_shifts.setdefault(id(inp), set()).add(offsets)
```

Comment (havogt) [reply src/gt4py/next/iterator/transforms/trace_shifts.py:39]: I thought the register mechanism is not nice, but it slightly changes the result: not in the result means no offset applied.

Comment (tehrengruber) [reply src/gt4py/next/iterator/transforms/trace_shifts.py:39]: As discussed TODO test case

File: tests/next_tests/unit_tests/iterator_tests/transforms_tests/test_trace_shifts.py [hunk 2]
```diff
@@ -165,13 +159,52 @@ def test_tuple_get():
         output=ir.SymRef(id="out"),
         domain=ir.FunCall(fun=ir.SymRef(id="cartesian_domain"), args=[]),
     )
-    expected = {"inp1": [], "inp2": [()]}  # never derefed  # once derefed
+    expected = {"inp1": set(), "inp2": {()}}  # never derefed  # once derefed
 
-    actual = dict()
-    TraceShifts().visit(testee, shifts=actual)
+    actual = TraceShifts.apply(testee)
     assert actual == expected
```
Comment (havogt) [inline tests/next_tests/unit_tests/iterator_tests/transforms_tests/test_trace_shifts.py:166]: a testcase seems to miss where we have an expression that is not derefed.

**Comment by havogt** [comment]
I'll approve once the testcase is added.

**Review by havogt — APPROVED** [review]

**Pull request merged**